go/libraries/doltcore/sqle/rowexec/lookup_join.go

```diff
@@ -0,0 +1,327 @@
+	mapping *lookupMapping,
+	joiner *prollyToSqlJoiner,
+	srcFilter, dstFilter, joinFilter sql.Expression,
+	isLeftJoin bool,
+	excludeNulls bool,
+) (*lookupJoinKvIter, error) {
+	if lit, ok := joinFilter.(*expression.Literal); ok {
+		if lit.Value() == true {
+			joinFilter = nil
+		}
+	}
+
+	return &lookupJoinKvIter{
+		srcIter:        srcIter,
+		dstIter:        targetIter,
+		joiner:         joiner,
+		keyTupleMapper: mapping,
+		srcFilter:      srcFilter,
+		dstFilter:      dstFilter,
+		joinFilter:     joinFilter,
+		isLeftJoin:     isLeftJoin,
+		excludeNulls:   excludeNulls,
+	}, nil
+}
+
+func (l *lookupJoinKvIter) Next(ctx *sql.Context) (sql.Row, error) {
+	for {
+		var err error
+		if l.dstKey == nil {
+			l.returnedARow = false
+			// get row from |src|
+			l.srcKey, l.srcVal, err = l.srcIter.Next(ctx)
+			if err != nil {
+				return nil, err
+			}
+			if l.srcKey == nil {
+				return nil, io.EOF
+			}
+
+			l.dstKey, err = l.keyTupleMapper.dstKeyTuple(ctx, l.srcKey, l.srcVal)
+			if err != nil {
+				return nil, err
+			}
+
+			if err := l.dstIter.New(ctx, l.dstKey); err != nil {
+				return nil, err
+			}
+		}
+
+		dstKey, dstVal, ok, err := l.dstIter.Next(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		if !ok {
+			l.dstKey = nil
+			if !(l.isLeftJoin && !l.returnedARow) {
+				continue
+			}
+		}
+
+		ret, err := l.joiner.buildRow(ctx, l.srcKey, l.srcVal, dstKey, dstVal)
+		if err != nil {
+			return nil, err
+		}
+
+		// side-specific filters are currently hoisted
+		if l.srcFilter != nil {
+			res, err := sql.EvaluateCondition(ctx, l.srcFilter, ret[:l.joiner.kvSplits[0]])
+			if err != nil {
+				return nil, err
+			}
+
+			if !sql.IsTrue(res) {
+				continue
+			}
+
+		}
+		if l.dstFilter != nil && l.dstKey != nil {
+			res, err := sql.EvaluateCondition(ctx, l.dstFilter, ret[l.joiner.kvSplits[0]:])
+			if err != nil {
+				return nil, err
+			}
+
+			if !sql.IsTrue(res) {
+				continue
+			}
+		}
+		if l.joinFilter != nil {
+			res, err := sql.EvaluateCondition(ctx, l.joinFilter, ret)
+			if err != nil {
+				return nil, err
+			}
+			if res == nil && l.excludeNulls {
+				// override default left join behavior
+				l.dstKey = nil
+				continue
+			} else if !sql.IsTrue(res) && l.dstKey != nil {
+				continue
+			}
+		}
+		l.returnedARow = true
+		return ret, nil
+	}
+}
+
+// lookupMapping is responsible for generating keys for lookups into
+// the destination iterator.
+type lookupMapping struct {
+	split      int
+	srcMapping val.OrdinalMapping
+
+	// litTuple are the statically provided literal expressions in the key expression
+	litTuple val.Tuple
+
+	litKd    val.TupleDesc
+	srcKd    val.TupleDesc
+	srcVd    val.TupleDesc
+	targetKb *val.TupleBuilder
+
+	ns   tree.NodeStore
+	pool pool.BuffPool
+}
+
+func newLookupKeyMapping(ctx context.Context, sourceSch schema.Schema, src prolly.Map, tgtKeyDesc val.TupleDesc, keyExprs []sql.Expression) *lookupMapping {
+	keyless := schema.IsKeyless(sourceSch)
+	var split int
```

Comment:
Add a comment explaining what `split` means?

Reply:
So like, it's the index into the schema separating the key and value fields? Can we make that clearer?